### PR TITLE
Add preview feature to links for ANET objects

### LINKED ISSUE #598: Improve LinkTo
I suggest to make:

- [ ] rename LinkTo to something more appropriate
- [ ] evaluate the default behavior for `whenUnspecified`, and where it is appropriate throughout ANET
- [ ] make LinkTo to be able to render content based on some hints. output types should be:
  - [x] interactive output. When rendering a person for example, provide background/rounded box, possibly an icon (the picture), and a preview on mouse hover
  - [x] minimal output. Same as above minus links, preview and with more sober styling
  - [ ] svg output. provide SVG nodes to be used by d3
- [ ] Add a preview ability for LinkTo

### PR DESCRIPTION
**Pull request by cemalettin-work**
Users will be able to hover over a link that links to an ANET object to see its preview in a popup window for a brief amount of time.

Closes #598
Closes [AB#208](https://dev.azure.com/ncia-anet/2aa083a5-af3d-44e1-8c7b-6e9e6b124d91/_workitems/edit/208)
Closes [AB#249](https://dev.azure.com/ncia-anet/2aa083a5-af3d-44e1-8c7b-6e9e6b124d91/_workitems/edit/249)
Closes [AB#247](https://dev.azure.com/ncia-anet/2aa083a5-af3d-44e1-8c7b-6e9e6b124d91/_workitems/edit/247)
Closes [AB#281](https://dev.azure.com/ncia-anet/2aa083a5-af3d-44e1-8c7b-6e9e6b124d91/_workitems/edit/281)

#### User changes
- Users will see a preview popup when they hover over some links.
- Advanced select components are going to display a preview when user hover over the options.

#### Super User changes
- none

#### Admin changes
- none

#### System admin changes
- [ ] anet.yml or anet-dictionary.yml needs change
- [ ] db needs migration
- [ ] documentation has changed
- [ ] graphql schema has changed

### Checklist
  - [x] Described the user behavior in PR body
  - [x] Referenced/updated all related issues
  - [x] commits follow a `repo#issue: Title` title format and [these 7 rules](https://chris.beams.io/posts/git-commit/)
  - [x] commits have a [clean history](https://epage.github.io/dev/commits/), otherwise PR may be squash-merged
  - [ ] Added and/or updated unit tests
  - [x] Added and/or updated e2e tests
  - [ ] Added and/or updated data migrations
  - [ ] Updated documentation
  - [x] Resolved all build errors and warnings
  - [x] Opened debt issues for anything not resolved here

### CODE REVIEW

File: client/src/components/AssignPositionModal.js [hunk 1]
```diff
@@ -99,8 +99,8 @@ const AssignPositionModal = ({ person, showModal, onCancel, onSuccess }) => {
       const errorMessage = (
         <>
           This position is currently held by{" "}
-          <LinkTo modelType="Person" model={position.person} />. By selecting
-          this position, they will be removed.
+          <LinkToNotPreviewed modelType="Person" model={position.person} />. By
```
Comment (ysf-simsoft) [inline client/src/components/AssignPositionModal.js:102]: When clicked on this link,  "Change Position Dialog" is automatically opened in the navigated page. Is this the desired behavior? Also, is it possible to display preview here? Wouldn't it make sense?

File: client/src/components/PositionTable.js [hunk 2]
```diff
@@ -57,7 +56,8 @@ const PositionTable = props => {
 
 PositionTable.propTypes = {
   // query variables for positions, when query & pagination wanted:
-  queryParams: PropTypes.object
+  queryParams: PropTypes.object,
+  linkToComp: PropTypes.func.isRequired
```
Comment (ysf-simsoft) [inline client/src/components/PositionTable.js:60]: PositionTable is used in 9 places and in 7 of them the value of `linkToComp` prop is `LinkTo`. So why not provide a default prop for that?

**Review by ysf-simsoft — CHANGES_REQUESTED** [review]

**Review by ysf-simsoft — CHANGES_REQUESTED** [review]

File: client/src/components/RelatedObjectNotes.js [hunk 3]
```diff
@@ -262,7 +262,7 @@ const RelatedObjectNotes = ({
               isJson && note.text ? utils.parseJsonSafe(note.text) : {}
```
[the remark this answers is not in the document]
Comment (ysf-simsoft) [reply client/src/components/RelatedObjectNotes.js:262]: Opened [a bug report](https://github.com/NCI-Agency/anet/issues/3355) for this. I am resolving this conversation.

File: client/src/components/AvatarDisplayComponent.js [hunk 4]
```diff
@@ -14,6 +21,7 @@ const AvatarDisplayComponent = ({ avatar, height, width, style }) => (
 
 AvatarDisplayComponent.propTypes = {
   avatar: PropTypes.string,
+  className: PropTypes.string,
```
[the remark this answers is not in the document]
Comment (cemalettin-simsoft) [reply client/src/components/AvatarDisplayComponent.js:24]: I am using className to target it in the preview mode, see https://github.com/NCI-Agency/anet/blob/GH-598-add-preview-to-LinkTo/client/src/components/ModelPreview.css
```
 /* center the avatar */
.linkto-model-preview-portal .model-preview .large-person-avatar{....
```

File: client/src/components/Kanban.js [hunk 5]
```diff
@@ -170,7 +171,7 @@ const Card = ({ task }) => {
                   <td>
                     {" "}
                     {this.props.task.taskedOrganizations.map(org => (
```
[the remark this answers is not in the document]
Comment (cemalettin-simsoft) [reply client/src/components/Kanban.js:173]: Hmm yes, probably refactored from class component and forgotten to remove `this`

File: client/src/components/CustomFields.js [hunk 6]
```diff
@@ -710,6 +756,7 @@ CustomFieldsContainer.propTypes = {
   fieldsConfig: PropTypes.object,
   formikProps: PropTypes.object,
   parentFieldName: PropTypes.string.isRequired,
+  linkToComp: PropTypes.func.isRequired,
```
[the remark this answers is not in the document]
Comment (cemalettin-simsoft) [reply client/src/components/CustomFields.js:759]: It is here because of circular dependency, there were 2 options, 

1- Move readonly components to a different file so only modify ReadonlyCustomFields props
2- Keep everything in the same file and change all LinkTo components to prop.

File: client/src/components/PositionTable.js (hunk 2 above)
Comment (cemalettin-simsoft) [reply client/src/components/PositionTable.js:60]: This is due to avoid circular dependency, we can't import `LinkTo` in this file.

File: client/src/components/AssignPositionModal.js (hunk 1 above)
Comment (cemalettin-simsoft) [reply client/src/components/AssignPositionModal.js:102]: Hmm, yes it might have been a copy paste mistake, this can and should show preview. Other problem is related to modal state not being closed even if we change person pages. Will fix them both

File: client/src/components/previews/OrganizationPreview.js [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+                            <i>No super users</i>
+                          </p>
+                        )}
+                      </>
+                    }
+                  />
+                )}
+
+                {organization.childrenOrgs &&
+                  organization.childrenOrgs.length > 0 && (
```
Comment (ysf-simsoft) [inline client/src/components/previews/OrganizationPreview.js:191]: Could use optional chaining here.

File: client/src/components/previews/ReportPreview.js [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+    if (error) {
+      return <p>Could not load the preview</p>
+    }
+    return null
+  }
+
+  let report
+
+  data.report.cancelled = !!data.report.cancelledReason
+  data.report.reportTags = (data.report.tags || []).map(tag => ({
```
Comment (ysf-simsoft) [inline client/src/components/previews/ReportPreview.js:95]: `report.tags` doesn't exist in gql.

File: client/src/components/previews/ReportPreview.js [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+    id: tag.uuid.toString(),
+    text: tag.name
+  }))
+  data.report.tasks = Task.fromArray(data.report.tasks)
```
Comment (ysf-simsoft) [inline client/src/components/previews/ReportPreview.js:99]: `report.tasks` doesn't exist in gql.

File: client/src/components/previews/ReportPreview.js [hunk 10]
```diff
@@ -0,0 +1,335 @@
...
+  data.report.reportPeople = Person.fromArray(data.report.reportPeople)
+  data.report.to = ""
```
Comment (ysf-simsoft) [inline client/src/components/previews/ReportPreview.js:101]: Is `report.to` a valid field?

**Review by ysf-simsoft — CHANGES_REQUESTED** [review]

File: client/src/components/previews/ReportPreview.js (hunk 8 above)
Comment (cemalettin-simsoft) [reply client/src/components/previews/ReportPreview.js]: Hmm yes, it might be better if the tasks are shown in the preview

File: client/src/components/previews/ReportPreview.js (hunk 10 above)
Comment (cemalettin-simsoft) [reply client/src/components/previews/ReportPreview.js:112]: It is used for emailing, not needed indeed.

**Review by ysf-simsoft — APPROVED** [review]

File: .github/renovate.json [hunk 11]
```diff
@@ -4,22 +4,32 @@
...
     "group:linters"
   ],
-  "lockFileMaintenance": { "enabled": true },
-  "labels": ["dependencies"],
+  "lockFileMaintenance": {
+    "enabled": true
+  },
+  "labels": [
+    "dependencies"
+  ],
```
Comment (gjvoosten) [inline .github/renovate.json:12]: All changes to this file seem to be unrelated to the actual issue, and there's also no separate commit explaining why they're here (they seem to have been introduced through a merge of the candidate branch). Please revert.

File: client/src/components/LinkToPreviewed.js [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+  showAvatar,
+  whenUnspecified,
+  className,
+  modelType,
+  model,
+  style,
+  previewId,
+  ...componentProps
+}) => {
+  componentProps.className = className
```
Comment (gjvoosten) [inline client/src/components/LinkToPreviewed.js:25]: I would not assign this here, but add `className={className}` to the props of `<LinkToComponent />` below.

File: client/src/components/LinkTo.js [hunk 13]
```diff
@@ -22,6 +22,10 @@ const LinkTo = ({
   style,
   ...componentProps
 }) => {
+  if (componentProps.previewId) {
+    // Previewed version consumes the previewId, so remove from this one
+    delete componentProps.previewId
+  }
```
Comment (gjvoosten) [inline client/src/components/LinkTo.js:28]: Since we're not interested in the return value of `delete`, might as well remove the `if (…)` and call `delete` unconditionally.

File: client/src/components/RichTextEditor.css [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+  font-weight: 300;
+}
+
+.anet .Draftail-Editor--focus .Draftail-Toolbar {
+  top: -15px;
+}
+
+.anet .Draftail-block--unstyled {
+  margin: 0.75rem 0;
+}
```
Comment (gjvoosten) [inline client/src/components/RichTextEditor.css:18]: This file should remain deleted; it is no longer used.

File: client/src/components/SimilarObjectsModal.js [hunk 15]
```diff
@@ -27,7 +28,10 @@ const SimilarObjectsModal = ({ objectType, userInput, onCancel }) => {
         <Modal.Title>Possible Duplicates</Modal.Title>
       </Modal.Header>
       <Modal.Body>
-        <TableComponent queryParams={queryParams} />
+        <TableComponent
+          queryParams={queryParams}
+          linkToComp={objectType === "Position" ? LinkToPreviewed : undefined}
```
Comment (gjvoosten) [inline client/src/components/SimilarObjectsModal.js:33]: This looks odd. Why is Position special here?

**Review by gjvoosten — CHANGES_REQUESTED** [review]

File: client/src/components/LinkToPreviewed.js [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+import ModelPreview from "components/ModelPreview"
+import ModelTooltip from "components/ModelTooltip"
+import _isEmpty from "lodash/isEmpty"
+import * as Models from "models"
+import PropTypes from "prop-types"
+import React from "react"
+import { Link } from "react-router-dom"
+
+const LinkToPreviewed = ({
+  componentClass,
```
Comment (gjvoosten) [inline client/src/components/LinkToPreviewed.js:13]: Is this prop _ever_ passed to this component?

**Review by gjvoosten — APPROVED** [review]
